qubesadmin/tools/qvm_template.py

```diff
@@ -0,0 +1,409 @@
+            tpl.features['template-release'] = \
+                package_hdr[rpm.RPMTAG_RELEASE]
+            tpl.features['template-install-date'] = \
+                str(datetime.datetime.today())
+            tpl.features['template-name'] = name
+            # TODO: Store source repo
+
+def qrexec_popen(args, app, service, stdout=subprocess.PIPE, encoding='UTF-8'):
+    if args.updatevm:
+        from_vm = shutil.which('qrexec-client-vm') is not None
+        if from_vm:
+            return subprocess.Popen(
+                ['qrexec-client-vm', args.updatevm, service],
+                stdin=subprocess.PIPE,
+                stdout=stdout,
+                stderr=subprocess.PIPE,
+                encoding=encoding)
+        else:
+            return subprocess.Popen([
+                    'qrexec-client',
+                    '-d',
+                    args.updatevm,
+                    'user:/etc/qubes-rpc/%s' % service
+                ],
+                stdin=subprocess.PIPE,
+                stdout=stdout,
+                stderr=subprocess.PIPE,
+                encoding=encoding)
+    else:
+        return subprocess.Popen([
+                '/etc/qubes-rpc/%s' % service,
+            ],
+            stdin=subprocess.PIPE,
+            stdout=stdout,
+            stderr=subprocess.PIPE,
+            encoding=encoding)
+
+def qrexec_payload(args, app, spec):
+    payload = ''
+    for r in args.enablerepo if args.enablerepo else []:
+        payload += '--enablerepo=%s\n' % r
+    for r in args.disablerepo if args.disablerepo else []:
+        payload += '--disablerepo=%s\n' % r
+    for r in args.repoid if args.repoid else []:
+        payload += '--repoid=%s\n' % r
+    payload += '--releasever=%s\n' % args.releasever
+    payload += spec + '\n'
+    payload += '---\n'
+    for fn in args.repo_files:
+        with open(fn, 'r') as f:
+            payload += f.read() + '\n'
+    return payload
+
+def qrexec_repoquery(args, app, spec='*'):
+    proc = qrexec_popen(args, app, 'qubes.TemplateSearch')
+    payload = qrexec_payload(args, app, spec)
+    stdout, stderr = proc.communicate(payload)
+    if proc.wait() != 0:
+        return None
+    result = []
+    for line in stdout.strip().split('\n'):
+        entry = line.split(':')
+        if not entry[0].startswith(PACKAGE_NAME_PREFIX):
+            continue
+        entry[0] = entry[0][len(PACKAGE_NAME_PREFIX):]
+        result.append(entry)
+    return result
+
+def qrexec_download(args, app, spec, path):
+    with open(path, 'wb') as f:
+        proc = qrexec_popen(args, app, 'qubes.TemplateDownload',
+            stdout=f, encoding=None)
+        payload = qrexec_payload(args, app, spec)
+        proc.stdin.write(payload.encode('UTF-8'))
+        proc.stdin.close()
+        while True:
+            c = proc.stderr.read(1)
+            if not c:
+                break
+            # Write raw byte w/o decoding
+            sys.stdout.buffer.write(c)
```

Comment:
The reason for this and the byte-by-byte reads below are because I'm currently reusing curl's progress indicator. That being said, this is indeed an issue for security. Maybe I should either implement my own progress indicator or pull in pv as a dependency?

Reply:
Implemented a simple progress bar for the time being.